### PR TITLE
[Enhancement] Battery Notification +Error Handlers

### PR DESCRIPTION
**Pull request by kRHYME7**
Continuation for  #414 
Discussion on Issue #411 

+ Non-intrusive Notification :bow:
+ Acts like the old script but prompts every 5% instead of per 5 minutes. (Upon reaching threshold) Stops if interval < 5.
+ Appending "on" "--dock" will Give Plug In/Out/Full
+ TUI
+ Error handling report on terminal 
+ Default for Battery full notification interval is set to 1 day.
+ Verbose mode for Debugging. 
+ Defaults on Screenshot. 
+ Should Fix If Device reports "Not" "Not charging" "Not Charging" or "Not"*

-- The script is too elaborate and could/should be refactored further. 

image

### CODE REVIEW

**Comment by devCKVargas** [comment]
Will test this. Just came back from vacation.
I'm liking the non-intrusive notif already.

**Comment by kRHYME7** [comment]
> I'm liking the non-intrusive notif already.
I'm guilty! Just woke up this morning seeing the issue posted on reddit instead of here in the repo. 

Already tried this on 2 different laptops plugging and unplugging my device just to replicate some errors lol.  

> Will test this. Just came back from vacation.

I't be much of hellp

Thanks !

File: Configs/.config/hypr/scripts/batterynotify.sh [hunk 1]
```diff
@@ -93,7 +93,7 @@ case "$battery_status" in         # Handle the power supply status
                     if [[ $battery_status != "Discharging" ]]; then
                     now=$(date +%s) 
                     if [[ "$prev_status" == *"Charging"* ]] || ((now - lt >= $((notify*60)) )); then
-                     fn_notify "-t 5000 -r 10" "CRITICAL" "Battery Full" "Please unplug your Charger"
+                     fn_notify "-t 5000 -r 69" "CRITICAL" "Battery Full" "Please unplug your Charger"
```
Comment (devCKVargas) [inline Configs/.config/hypr/scripts/batterynotify.sh:96]: Haha Nice!
But I'd much rather have my laptop charged to either 60 or 80%

File: Configs/.config/hypr/scripts/batterynotify.sh [hunk 2]
```diff
@@ -19,8 +19,8 @@ while (( "$#" )); do  # Parse command-line arguments and defaults
       echo "  --low, -l         Set battery low threshold (default: $mnl% percent)"
       echo "  --unplug, -u      Set unplug charger threshold (default: $mxu% percent )"
       echo "  --timer, -t       Set countdown timer (default: $mnt seconds)"
-      echo "  --interval, -i    Set notify interval  on LOW UNPLUG Status  (default: $mni% percent)"
-      echo "  --notify, -n      Set notify interval for Battery Full Status  (default: $mnn minutes)"
+      echo "  --interval, -i    Set notify interval  on LOW UNPLUG Status  (default: 2% percent)"
```
Comment (devCKVargas) [inline Configs/.config/hypr/scripts/batterynotify.sh:22]: 2% is a very critical battery level. especially on old laptops, they'd just power off. Some battery would start to count down in seconds alongside their current capacity eg. starting from 10% and would black screen before they even reached that low capacity.
I'd suggest the Low battery would start at 20% with <x time> interval notification. & critical would maybe at maybe 5% or 10% so user would have time to plug it in.

File: Configs/.config/hypr/scripts/batterynotify.sh [hunk 3]
```diff
@@ -1,107 +1,149 @@
...
-                            count=$((count-1))
-                            sleep 1
-                        done
-                        if [ $count -eq 0 ]; then
-                             handle_action
-                        fi
-                    elif [[ "$battery_percentage" -le "$battery_low_threshold" ]] && (( (last_notified_percentage - battery_percentage) >= 1 )); then
-                        send_notification "-r 10" "CRITICAL" "Battery Low" "Battery is at $battery_percentage%. Connect the charger."
-                        last_notified_percentage=$battery_percentage
+                        fn_notify   "-t 5000 -r 10" "$urgency" "Charger Plug OUT" "Battery is at $battery_percentage%."
```
Comment (devCKVargas) [inline Configs/.config/hypr/scripts/batterynotify.sh:73]: Let's change this to Plugged*

File: Configs/.config/hypr/scripts/batterynotify.sh [hunk 4]
```diff
@@ -1,107 +1,149 @@
...
+                    fn_percentage 
                     ;;
-                "Charging")                     
-                    if [[ "$prev_status" == "Discharging" ]]; then
+                "Not"*|"Charging") # Due to modifications of some devices Not Charging after reaching 99 or limits
+                    if [[ ! -f "/tmp/hyprdots.batterynotify.status.$battery_status-$$" ]] && [[ "$battery_status" == "Not"* ]] ; then 
+                    touch "/tmp/hyprdots.batterynotify.status.$battery_status-$$"
+                    count=$(( timer > $mnt ? timer :  $mnt )) # reset count                    
+                    echo "Status: '==>> "$battery_status" <<==' Device Reports Not Charging!,This may be device Specific errors."
+                    fn_notify  "-t 5000 -r 10" "CRITICAL" "Charger Plug In" "Battery is at $battery_percentage%."
```
Comment (devCKVargas) [inline Configs/.config/hypr/scripts/batterynotify.sh:82]: This one as well

**Comment by devCKVargas** [comment]
The notification is hit or miss. Takes maybe about 10s, sometimes 5, sometimes it won't even show itself when the charging state changed.

**Comment by kRHYME7** [comment]
> The notification is hit or miss. Takes maybe about 10s, sometimes 5, sometimes it won't even show itself when the charging state changed.





Thanks, I have  3 inferences why this happens.

1st. Multiple events happen at once, making notification overload. Maybe I should solve this.

2nd. Dbus-monitor not caching yet needs multiple reboots maybe. Just speculation.

**Comment by kRHYME7** [comment]
@devCKVargas  Hello I have added some verbose things can you run 
 ```  
~/.config/hypr/scripts/batterynotify.sh -v 
```

**Comment by oxzh** [comment]
now is infinity ping, 
image
 not even new notification remove them 
image

**Comment by T-Crypt** [comment]
I also pulled the latest batterynotify.sh and it keeps yelling at me. I have my laptop docked -- I liked the old script that we had, it almost never yelled at me. 

May need to push the old battery notify and then draft this PR and then test on a laptop before merging this.

**Comment by kRHYME7** [comment]
Thank You @T-Crypt  @devCKVargas  @oxzh for the Reviews.<3

Will you try my last commit? If it is still intrusive I'll probably pull the old script.

This latest commit should act like the old script  unless appended by "on" or "--undock"

Thank you for baring this messy work. :bowing_man:

**Comment by prasanthrangan** [comment]
@kRHYME7 @devCKVargas @oxzh @T-Crypt

Thank you so much guys for your time and effort to develop and test this!
Could someone please let me know if this script works as expected and ready to merge? or are you guys still improving it?

**Comment by kRHYME7** [comment]
It should work like > Acts like the old script but prompts every 5% instead of per 5 minutes. (Upon reaching threshold) Stops if interval < 5.

Thank you everyone!

**Review by prasanthrangan — APPROVED** [review]
cool thanks, please make sure to keep the script as light as possible, so that when its run as daemon it wont drain system resources...

**Comment by kRHYME7** [comment]
> cool thanks, please make sure to keep the script as light as possible, so that when its run as daemon it won't drain system resources...

Aye aye Captain. The process should be very light because it will only execute something if dbus-monitor tickles the loop. I'll still monitor and refactor this.

**Pull request merged**